### PR TITLE
More accurate spans for arg removal suggestion

### PR DESCRIPTION
**Pull request by estebank**
Partially address #106304.

### CODE REVIEW

File: src/test/ui/argument-suggestions/basic.stderr [hunk 1]
```diff
@@ -25,8 +25,9 @@ LL | fn extra() {}
    |    ^^^^^
 help: remove the extra argument
    |
-LL |     extra();
-   |          ~~
+LL -     extra("");
```
Comment (fmease) [inline src/test/ui/argument-suggestions/basic.stderr:28]: What do think about shrinking the replacement span in those simple cases where the diff is (- line, +line) to reduce verbosity?

**Comment by petrochenkov** [comment]
r? rust-lang/compiler
I'd like to take some break from reviewing.

**Comment by estebank** [comment]
@petrochenkov Consider removing yourself from the review rotation and readding yourself once you are ready to take them on, just to avoid spam on your end :)

Happy new year.

**Comment by Ezrashaw** [comment]
@estebank I have some code for the case where arguments must be added, but that code relies on this code to be merged first.

Also I think I have found a bug with this PR:
```rust
fn foo(x: &str) {}

fn main() {
    foo(1, 2, "hello");
}
```
emits this:
```text
error[E0061]: this function takes 1 argument but 3 arguments were supplied
 --> <anon>:7:5
  |
7 |     foo(1, 2, "hello");
  |     ^^^ -  - argument of type `{integer}` unexpected
  |         |
  |         argument of type `{integer}` unexpected
  |
note: function defined here
 --> <anon>:2:4
  |
2 | fn foo(x: &str) {
  |    ^^^ -------
help: remove the extra arguments
  |
7 -     foo(1, 2, "hello");
7 +     foo(, "hello");
  |
```

Note the leading comma.

**Comment by estebank** [comment]
Yeah, you're right :-/

Would you have time to try and fix the bug in a subsequent PR?

**Comment by Ezrashaw** [comment]
Yeah, I can have a look at this. Honestly the whole error printing bit of that function needs a rewrite, it's all based around the whole-span idea. Ideally we can get this merged soon so that I can create my PRs for the additional suggestion and to fix that bug.

**Comment by Ezrashaw** [comment]
@estebank What's happening with this PR, I intend to write some more changes but obviously they rely on this PR. Also could you look at my review comment, thanks!

**Comment by estebank** [comment]
@Ezrashaw this PR is awaiting review. Different reviewers have different schedules when they are available.

File: compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs [hunk 2]
```diff
@@ -1081,6 +1121,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                prev = provided_idx.index() as i64;
+            }
+            let idx = ProvidedIdx::from_usize((prev + 1) as usize);
+            if let None = provided_idx
+                && let Some((_, arg_span)) = provided_arg_tys.get(idx)
+            {
+                let (_, expected_ty) = formal_and_expected_inputs[expected_idx];
+                suggestions.push((*arg_span, ty_to_snippet(expected_ty, expected_idx)));
+            }
+        }
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1137]: What do you mean by  "breaking your code"? Do you have an example of a broken suggestion that we can put as a test?

estebank: I don't really understand what this is doing either.

File: compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs [hunk 3]
```diff
@@ -1081,6 +1121,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+            if let Some(provided_idx) = provided_idx {
+                prev = provided_idx.index() as i64;
+            }
+            let idx = ProvidedIdx::from_usize((prev + 1) as usize);
+            if let None = provided_idx
+                && let Some((_, arg_span)) = provided_arg_tys.get(idx)
+            {
+                let (_, expected_ty) = formal_and_expected_inputs[expected_idx];
+                suggestions.push((*arg_span, ty_to_snippet(expected_ty, expected_idx)));
+            }
```
Comment (cjgillot) [inline compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1136]:
```suggestion
            if let Some(provided_idx) = provided_idx {
                prev = provided_idx.index() as i64;
            } else {
                let idx = ProvidedIdx::from_usize((prev + 1) as usize);
                if let Some((_, arg_span)) = provided_arg_tys.get(idx) {
                    let (_, expected_ty) = formal_and_expected_inputs[expected_idx];
                    suggestions.push((*arg_span, ty_to_snippet(expected_ty, expected_idx)));
                }
            }
```
?

File: src/test/ui/alloc-error/alloc-error-handler-bad-signature-3.stderr [hunk 4]
```diff
@@ -17,8 +17,10 @@ LL | fn oom() -> ! {
    = note: this error originates in the attribute macro `alloc_error_handler` (in Nightly builds, run with -Z macro-backtrace for more info)
 help: remove the extra argument
    |
-LL | fn oom() -> !() {
-   |              ++
+LL - fn oom() -> ! {
+LL -     loop {}
+LL - }
```
Comment (cjgillot) [inline src/test/ui/alloc-error/alloc-error-handler-bad-signature-3.stderr:22]: This suggestion is weird too. Any idea why?

File: compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs [hunk 5]
```diff
@@ -892,6 +917,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     };
                     labels
                         .push((provided_span, format!("argument{} unexpected", provided_ty_name)));
```
Comment (cjgillot) [inline compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:919]: Drive-by
```suggestion
                        .push((provided_span, format!("unexpected argument{}", provided_ty_name)));
```

File: src/test/ui/argument-suggestions/extra_arguments.stderr [hunk 6]
```diff
@@ -212,8 +224,9 @@ LL | fn two_arg_same(_a: i32, _b: i32) {}
    |    ^^^^^^^^^^^^ -------  -------
 help: remove the extra argument
    |
-LL |   two_arg_same(1, 1);
-   |               ~~~~~~
+LL -     1,
+LL +     1
+   |
```
[the remark this answers is not in the document]
Comment (estebank) [reply src/test/ui/argument-suggestions/extra_arguments.stderr:229]: I would like to address this in a separate PR. This is a consequence of how the suggestion machinery works. I find it unreasonable to address that as part of an "unrelated" change.

File: compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs [hunk 7]
```diff
@@ -1098,7 +1153,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 Some(format!("provide the argument{}", if plural { "s" } else { "" }))
             }
             SuggestionText::Remove(plural) => {
-                Some(format!("remove the extra argument{}", if plural { "s" } else { "" }))
+                err.multipart_suggestion_verbose(
```
[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1156]: I tend to prefer the verbose output in general, because I feel showing the resulting code ends up being clearer, but I can change this to terse, it shouldn't be too much of an issue.

File: src/test/ui/alloc-error/alloc-error-handler-bad-signature-3.stderr (hunk 4 above)
Comment (estebank) [reply src/test/ui/alloc-error/alloc-error-handler-bad-signature-3.stderr:22]: I think this is caused by the macro spans :-/

File: compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs]: Changed. It's on its own commit so that you can review it.

File: compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:1152]: It's been a while since I wrote this and now I don't recall *exactly* what this does, but this weird code was so that I could also incorporate the "change this argument for this other one" suggestions into the removal one, but it is quite messy :-/

[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs]: The last commit now shows the effect of this block taken out on the tests.

**Comment by TaKO8Ki** [comment]
Cloud you rebase the commits?

**Comment by estebank** [comment]
@TaKO8Ki I just rebased and pushed, but the changes were merging cleanly.
